sdk/core/allocator/alloc.h

```diff
@@ -1618,57 +2009,103 @@ class MState
 	 *
 	 * Note that this chunk does not have to come from quarantine, because it
 	 * can come from initialisation or splitting a free chunk.
+	 *
+	 * Initializes the linkages of p.
 	 */
 	void mspace_free_internal(MChunk *p)
 	{
 		ok_in_use_chunk(p);
-		size_t psize = 0, psizeold = 0;
-		if (RTCHECK(ok_address(p->ptr()) && p->ok_in_use()))
-		{
-			// Clear the shadow bits.
-			revoker.shadow_paint_range(
-			  chunk2mem(p).address(), p->chunk_next()->ptr(), false);
-			psize           = p->size_get();
-			psizeold        = psize;
-			MChunk *current = p;
-			MChunk *next    = p->chunk_next();
-			if (!p->is_prev_in_use())
+
+		// Clear the shadow bits.
+		revoker.shadow_paint_range(
+		  chunk2mem(p).address(), p->chunk_next()->ptr(), false);
+
+		heapFreeSize += p->size_get();
+
+		if (!p->is_prev_in_use())
+		{
+			// Consolidate backward
+			MChunk *prev = p->chunk_prev();
+			unlink_chunk(prev, prev->size_get());
+			ds::linked_list::unsafe_remove_link(&prev->header, &p->header);
+			p->header.clear();
+			p = prev;
+		}
+
+		MChunk *next = p->chunk_next();
```

Comment:
At this point, if `prev` is not in use, have we updated the size field to include both `p` and `prev` so that `p->chunk_next()` can work?

If not, I think we should move this line above `is_prev_in_use()` so that `chunk_next()` is valid.

Reply:
Yep